### ISSUE TITLE
Logging libraries

### ISSUE THREAD

**Issue by raigoinabox** [open-eid/digidoc4j#74]
Hi,

In my companies product, we recently switched from log4j to logback through slf4j. After making the switch, I would like to ensure, that there are no more references to log4j classes in our project now or, by accident, in the future. For that I would like to have the log4j classes not exist in the classpath.

However, the [digidoc4j](https://mvnrepository.com/artifact/org.digidoc4j/digidoc4j/3.3.0) library has [log4j-over-slf4j](https://mvnrepository.com/artifact/org.slf4j/log4j-over-slf4j/1.7.26) as a maven dependency and that library does put the log4j classes in the classpath. I would like you to remove that dependency or make it optional.

In addition the [digidoc4j](https://mvnrepository.com/artifact/org.digidoc4j/digidoc4j/3.3.0) library pulls in [logback-classic](https://mvnrepository.com/artifact/ch.qos.logback/logback-classic/1.2.3) and [logback-core](https://mvnrepository.com/artifact/ch.qos.logback/logback-core/1.2.3) as maven dependencies. For now, we are on logback, but if we want to in the future use another logging library (like log4j2) then [digidoc4j](https://mvnrepository.com/artifact/org.digidoc4j/digidoc4j/3.3.0) is currently blocking that path or will make it harder to migrate. I would like you to remove the [logback-classic](https://mvnrepository.com/artifact/ch.qos.logback/logback-classic/1.2.3) and [logback-core](https://mvnrepository.com/artifact/ch.qos.logback/logback-core/1.2.3) dependencies as well or make them optional.

I would like to also mention that even though logback is better, the log4j to logback migration was pushed onto us by the [digidoc4j](https://mvnrepository.com/artifact/org.digidoc4j/digidoc4j/3.3.0) library completely suddenly when we weren't ready for it and caused us a few headaches in addition to the work necessary for a migration.

In summary, please remove the [log4j-over-slf4j](https://mvnrepository.com/artifact/org.slf4j/log4j-over-slf4j/1.7.26), [logback-classic](https://mvnrepository.com/artifact/ch.qos.logback/logback-classic/1.2.3) and as maven dependencies.

**Comment by elion** [comment]
Have you tried excluding transitive dependencies manually?

It is important that the proposed changes do not break logback compatibility.

**Comment by raigoinabox** [comment]
I am wary of excluding the dependencies manually. I am worried that somewhere inside the digidoc4j library, there is either a direct import to either logging library class via a cast or a direct field declaration. This would cause a RuntimeException inside the application. Since the digidoc4j library is large, it would be hard to know if all the required corners are tested. The cost of an exception doesn't seem the benefit.

On the library side however, after removing them, on compile you would immediately get compile time errors on all potential error locations. Much safer to remove the dependency here.

Can you explain to me the backwards compatibility potential issues?

**Comment by naare** [comment]
We have removed the dependencies with this commit: https://github.com/open-eid/digidoc4j/commit/efe12f6cd229900cd66adb7a234e8f13e6ce8f65

It will be part of final 4.0.0 release. Integrators will need to add the prefferred logging dependencies.
